src/antworld/route_ardin.cc

```diff
@@ -241,16 +242,16 @@ void RouteArdin::load(const std::string &filename, bool realign)
     glVertexPointer(2, GL_FLOAT, 0, BUFFER_OFFSET(0));
     glEnableClientState(GL_VERTEX_ARRAY);
 
-    {
-        // Bind and upload zeros to colour buffer
-        std::vector<uint8_t> colours(m_Waypoints.size() * 3, 0);
-        glBindBuffer(GL_ARRAY_BUFFER, m_WaypointsColourVBO);
-        glBufferData(GL_ARRAY_BUFFER, m_Waypoints.size() * sizeof(uint8_t) * 3, colours.data(), GL_DYNAMIC_DRAW);
-
-        // Set colour pointer and enable client state in VAO
-        glColorPointer(3, GL_UNSIGNED_BYTE, 0, BUFFER_OFFSET(0));
-        glEnableClientState(GL_COLOR_ARRAY);
-    }
+    // Resize and zero colours
+    m_Colours.resize(m_Waypoints.size(), {0, 0, 0});
+
+    // Bind and upload colour buffer
+    glBindBuffer(GL_ARRAY_BUFFER, m_WaypointsColourVBO);
+    glBufferData(GL_ARRAY_BUFFER, m_Colours.size() * sizeof(uint8_t) * 3, m_Colours.data(), GL_DYNAMIC_DRAW);
```

Comment:
So the change here is basically to always store a CPU-side copy of the colours so you can highlight bits and re-upload them to the GPU without having to reallocate the CPU memory. Because OpenGL arrays are entirely flat i.e. it's an array of ``uint8_t`` and because, for convenience, ``m_Colours`` is a vector of ``std::array<uint8_t, 3>`` we need to multiply the size of the vector by 3 to get number of components.

Reply:
Ohh I see.